subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskDependencyInferenceIntegrationTest.groovy

```diff
@@ -0,0 +1,252 @@
+            tasks.register("b") {
+                dependsOn provider
+            }
+        """
+
+        when:
+        run("b")
+
+        then:
+        result.assertTasksExecuted(":a", ":b")
+    }
+
+    def "dependency declared using provider mapping that returns task implies dependency on task"() {
+        buildFile << """
+            def a = tasks.create("a")
+            def provider = provider { "a" }
+            tasks.register("b") {
+                dependsOn provider.map { tasks.getByName(it) }
+            }
+        """
+
+        when:
+        run("b")
+
+        then:
+        result.assertTasksExecuted(":a", ":b")
+    }
+
+    def "dependency declared using task output file property implies dependency on task"() {
+        taskTypeWithOutputFileProperty()
+        buildFile << """
+            def task = tasks.create("a", OutputFileTask) {
+                outFile = file("a.txt")
+            }
+            tasks.register("b") {
+                dependsOn task.outFile
+            }
+        """
+
+        when:
+        run("b")
+
+        then:
+        result.assertTasksExecuted(":a", ":b")
+    }
+
+    def "dependency declared using mapped task output file property implies dependency on task and does not run mapping function"() {
+        taskTypeWithOutputFileProperty()
+        buildFile << """
+            def task = tasks.create("a", OutputFileTask) {
+                outFile = file("a.txt")
+            }            
+            tasks.register("b") {
+                dependsOn task.outFile.map { throw new RuntimeException() }
+            }
+        """
+
+        when:
+        run("b")
+
+        then:
+        result.assertTasksExecuted(":a", ":b")
+    }
+
+    def "input file collection containing task provider implies dependency on all outputs of the task"() {
+        taskTypeWithMultipleOutputFiles()
+        taskTypeWithInputFilesProperty()
+        buildFile << """
+            def provider = tasks.register("a", OutputFilesTask) {
+                out1 = file("file1.txt")
+                out2 = file("file2.txt")
+            }
+            tasks.register("b", InputFilesTask) {
+                inFiles.from(provider)
+                outFile = file("out.txt")
+            }
+        """
+
+        when:
+        run("b")
+
+        then:
+        result.assertTasksExecuted(":a", ":b")
+        file("out.txt").text == "1,2"
+    }
+
+    def "input file collection containing mapped task provider implies dependency on a specific output of the task"() {
+        taskTypeWithMultipleOutputFiles()
+        taskTypeWithInputFilesProperty()
+        buildFile << """
+            def provider = tasks.register("a", OutputFilesTask) {
+                out1 = file("file1.txt")
+                out2 = file("file2.txt")
+            }
+            tasks.register("b", InputFilesTask) {
+                inFiles.from provider.map { project.layout.projectDir.file(it.out1.absolutePath) }
```

Comment:
What happens if, as a plugin developer, I don't use ` project.layout.projectDir.file(...` what if I instead just use `provider.map { it.out1 }`?